spec-0007/index.md

```diff
@@ -0,0 +1,183 @@
+2. If `rng=None`, handle `random_state` as in legacy behavior (see above), except use a compatible Generator instead of RandomState.
+   A DeprecationWarning is raised to warn about a future change in behavior.
+3. After <X time>, use only `rng`, seeding with `default_rng(rng)`.
+   Raise an error if `random_state` is provided.
+4. At a time of the library's choosing, remove any machinery related to `random_state`.
+
+### Impact
+
+The following users will be affected:
+
+1. Those who use `np.random.seed`. The proposal will do away with that global seeding mechanism, meaning that code that relies on it will, after a certain deprecation period, start seeing a different stream of random numbers than before. To ensure that this does not go unnoticed, the library should raise a `FutureWarning` if `np.random.seed` was called earlier (we show how to do that further down).
+
+   Such code will, in effect, go from being seeded to being unseeded.
+   To avoid that from happening, the code will have to be modified to pass in explicitly an `rng` argument on each function call.
+
+2. Those who do not seed. Their code will, after the deprecation period, use the newly proposed default. Since they were already not requesting repeatable sequences, and since the underlying _distributions_ of pseudo-random numbers did not change, they should be unaffected.
+
+3. Users of `random_state=...`. Support for the `random_state` argument may be dropped eventually, but meanwhile we can provide clear guidance, via deprecation warnings and documentation, on how to migrate to the new `rng` keyword.
+
+[^hinsen]: The Hinsen principle states, loosely, that code should, whether executed now or in the future, return the same result, or raise an error.
+
+### Code
+
+As an example, consider how SciPy would transition from the `seed` to the `rng` keyword using a decorator.
+This is implemented using:
+1. A `check_random_state` function which normalizes either old (`seed`) or new (`rng`) input to a `Generator` object.
+   If neither `seed` nor `rng` was passed but the user has previously called `np.random.seed()`
+   this function gives a `FutureWarning` because the behavior will change as noted in
+   the Impact section point 1.
+2. A decorator to deal with the `seed` to `rng` keyword rename.  In future versions, this will deprecate the `seed` keyword. Meanwhile, it ensures that the documentation and auto-completion only advertises the new parameter name.
+   Delaying the deprecation ensures that downstream users can switch to `rng=` on all supported SciPy versions when the deprecation happens.
+
+```python
+_NoValue = object()  # singleton to indicate not explicitly passed
+
+
+def check_random_state(seed=_NoValue, rng=_NoValue):
+    if rng is not _NoValue and seed is not _NoValue:
+        raise TypeError("cannot pass both `rng=` and `random_state=` at the same time.")
+    if rng is not _NoValue:
+        return np.random.default_rng(rng)
```

Comment:
IIRC, this will allow `rng` to be an integer, but below, it says 
> takes an `rng` argument which is guaranteed to be a either a Generator or a RandomState

If we want to train users to think about `rng`s and less about `seed`s, it might make sense to start by insisting on a RNG object and not allowing an integer seed.

Reply:
I don't recall we explicitly discussed this yet, good that you are bringing this up.

I am +1 on not allowing integers in the long run to, as you said, force people to think about what they are doing when they use a rng. Thinking about it, if you want reproducibility you probably want to seed multiple functions so why pass an integer everywhere instead of a Generator object.

Reply:
It was part of the discussion when we designed `np.random.default_rng()`. We wanted to have a replacement for third-party `check_random_state()` functions inside of `np.random` itself. SPEC7 is about coordinating the deprecation strategy for getting people to move from the pattern of `random_state = check_random_state(random_state)` to `rng = np.random.default_rng(rng)`, not really really re-envisioning much else. This pattern from `check_random_state()` (in it's various flavors) has been battle-tested since its original incarnation in `scikit-learn` for over a decade. Another possibility that was considered was having `default_rng()` only take seeds (or `None`) while also including a `check_rng(rng)` that only takes `Generator`s, `None`, and legacy `RandomState`s, but no one then had the appetite for multiple functions like this, given the success of `check_random_state()`. If you really want to relitigate the issue, the proper venue would be an enhancement proposal for numpy, not a SPEC.

In documentation, I do highly encourage emphasizing that we expect people to pass around `Generator`s primarily, but allowing for a seed (of various kinds) or legacy `RandomState`s are permitted for rarer cases. We don't always know what the entrypoint our user will use in an API. While creating a single `Generator` at the top of a program and passing it all the way through everything is frequently the right thing to do, especially in ML, you come across cases where you will need to rerun one thing (like a CV splitter) multiple times and have it do the same thing (possibly to multiple sequences) separate from the main PRNG stream. Allowing for straight static seeds helps with that (though I do admit there is little loss in requiring people to create the `Generator` from the seed separately).

Reply:
This SPEC is about what we want it to be about. Personally I am not putting much weight in the argument of time nor existing APIs, simply because a lot of past decisions were made with backward compatibility in mind. Here we have a real opportunity for change.

Let's see what others think.

Reply:
I'm not saying there shouldn't be a discussion. I'm saying that we want the One True Pattern to be in numpy, not the examples of a SPEC to be copied with variations in all the libraries. Right now, the blessed One True Pattern is `default_rng()`. If you want a `default_rng()/check_rng()` pair, propose it in numpy.

Reply:
Precisely, this is a SPEC and not a NEP. i.e. we can deviate from what NumPy says however we want.

Reply:
Still, I think having a reference implementation of any code feature that ships along with NumPy is probably wise. And no reason we shouldn't align support patterns between NumPy and the SPEC.

Reply:
My point is precisely that such a change *should* be a NEP and not a SPEC. What you are proposing is certainly plausible. But it was considered and rejected (and no, not for backwards compatibility reasons, but on the grounds of exactly the same information that we have available to us right now).

Reply:
OK, I think I understand that integers and such are still accepted because we want to be able to validate `rng` arguments simply by calling `np.random.default_rng(rng)`; we don't want to have additional checks to reject other types. 

It sounds like we agree that `Generators` should be encouraged, though, even if other types are accepted. With that in mind, how about this for suggested documentation:

```
    rng : `numpy.random.Generator`, optional

        Pseudorandom number generator state. For legacy reasons, types
        other than `numpy.random.Generator` are accepted and passed to
        `numpy.random.default_rng` to instantiate a `Generator`.
```

Reply:
I'd drop "For legacy reasons", but yeah, that works. Consider having a short sentence about `None` providing a fresh, independent `Generator`, since this is a main point of difference between `rng=None` and `random_state=None`; it can use the repetition (unlike all the other minutiae which can be left to the `default_rng` reference).

Reply:
```
rng : `numpy.random.Generator`, optional

    Pseudorandom number generator state.  When `rng` is None, a new
    `numpy.random.Generator` is created using entropy from the
    operating system. Types other than `numpy.random.Generator` are
    passed to `numpy.random.default_rng` to instantiate a `Generator`. 
```

Reply:
I included this in the example function docstring.